nucypher/acumen/perception.py

```diff
@@ -30,146 +33,323 @@
-            return
-        sorted_nodes = self.sorted()
-
-        sorted_nodes_joined = b"".join(bytes(n) for n in sorted_nodes)
-        checksum = keccak_digest(sorted_nodes_joined).hex()
-        if checksum not in self.states:
-            self.checksum = keccak_digest(b"".join(bytes(n) for n in self.sorted())).hex()
-            self.updated = maya.now()
-            # For now we store the sorted node list.  Someday we probably spin this out into
-            # its own class, FleetState, and use it as the basis for partial updates.
-            new_state = self.FleetState(nickname=self.nickname,
-                                        nodes=sorted_nodes,
-                                        icon=self.icon,
-                                        updated=self.updated,
-                                        checksum=self.checksum)
-            self.states[checksum] = new_state
-            return checksum, new_state
-
-    def start_tracking_state(self, additional_nodes_to_track=None):
-        if additional_nodes_to_track is None:
-            additional_nodes_to_track = list()
-        self.additional_nodes_to_track.extend(additional_nodes_to_track)
-        self._tracking = True
-        self.update_fleet_state()
-
-    def sorted(self):
-        nodes_to_consider = list(self._nodes.values()) + self.additional_nodes_to_track
-        return sorted(nodes_to_consider, key=lambda n: n.checksum_address)
+    def items(self):
+        return self._current_state.items()
 
-    def shuffled(self):
-        nodes_we_know_about = list(self._nodes.values())
-        random.shuffle(nodes_we_know_about)
-        return nodes_we_know_about
+    def values(self):
+        return self._current_state.values()
+
+    def latest_states(self, quantity):
+        """
+        Returns at most ``quantity`` latest archived states (including the current one),
+        in chronological order.
+        """
+        return self._archived_states[-min(len(self._archived_states), quantity):]
+
+    def addresses(self):
+        return self._current_state.addresses()
 
-    def abridged_states_dict(self):
-        abridged_states = {}
-        for k, v in self.states.items():
-            abridged_states[k] = self.abridged_state_details(v)
-        return abridged_states
+    def snapshot(self):
+        return self._current_state.snapshot()
 
     @staticmethod
-    def abridged_state_details(state):
-        return {"nickname": str(state.nickname),
-                # FIXME: generalize in case we want to extend the number of symbols in the state nickname
-                "symbol": state.nickname.characters[0].symbol,
-                "color_hex": state.nickname.characters[0].color_hex,
-                "color_name": state.nickname.characters[0].color_name,
-                "updated": state.updated.rfc2822(),
-                }
+    def unpack_snapshot(data):
+        return FleetState.unpack_snapshot(data)
+
+    def record_fleet_state(self):
```

Comment:
There will be some consequences to the status monitor code which will need to be updated - https://github.com/nucypher/nucypher-monitor/blob/master/monitor/crawler.py#L251. Not a huge deal, just something to note (@KPrasch )

Reply:
Thanks for noticing that, it'll have to be updated.